UI/States/CraftingUIState.cs

```diff
@@ -1206,30 +1206,57 @@ or ButtonConfigurationMode.LegacyWithGear
 						CraftingGUI.forceSpecificRecipeResort = true;
 					} else if (MagicStorageConfig.RecipeBlacklistEnabled && Main.keyState.IsKeyDown(Keys.LeftControl)) {
 						if (recipeButtons.Choice == CraftingGUI.RecipeButtonsBlacklistChoice) {
-							if (storagePlayer.HiddenRecipes.Remove(obj.StoredItem)) {
-								Main.NewText(Language.GetTextValue("Mods.MagicStorage.RecipeRevealed", Lang.GetItemNameValue(obj.StoredItem.type)));
-
-								slotZone.SetItemsAndContexts(int.MaxValue, GetRecipe);
-							}
+							Action<StoragePlayer, Item> revealRecipe = Main.keyState.IsKeyDown(Keys.LeftShift) ? RevealGlobalRecipe : RevealPlayerRecipe;
+							revealRecipe(storagePlayer, obj.StoredItem);
 						} else {
-							if (storagePlayer.HiddenRecipes.Add(obj.StoredItem)) {
-								Main.NewText(Language.GetTextValue("Mods.MagicStorage.RecipeHidden", Lang.GetItemNameValue(obj.StoredItem.type)));
-
-								slotZone.SetItemsAndContexts(int.MaxValue, GetRecipe);
-							}
+							Action<StoragePlayer, Item> hideRecipe = Main.keyState.IsKeyDown(Keys.LeftShift) ? HideGlobalRecipe : HidePlayerRecipe;
+							hideRecipe(storagePlayer, obj.StoredItem);
 						}
 
 						StorageGUI.SetRefresh();
 						CraftingGUI.SetNextDefaultRecipeCollectionToRefresh(obj.StoredItem.type);
 					} else {
 						CraftingGUI.SetSelectedRecipe(CraftingGUI.recipes[objSlot]);
 						(parentUI as CraftingUIState).history.AddHistory(CraftingGUI.selectedRecipe);
-						
+
 						StorageGUI.SetRefresh();
 
 						parentUI.UpdatePanelHeight(parentUI.PanelHeight);
 					}
 				};
+
+				void RevealPlayerRecipe(StoragePlayer storagePlayer, Item item) {
```

Comment:
Not a must, but consider moving the functions to inside the lambda since that would allow you to remove the `StoragePlayer` parameter

Reply:
Lambda is more expensive than local function, and I think local function is easier to understand.

Reply:
I think that what Exterminator was suggesting was the move the function logic into the lambda

Reply:
> I think that what Exterminator was suggesting was the move the function logic into the lambda

I know that using a lambda can expand the capture scope to one less parameter. It's just that I think the code will look uglier.

Reply:
What I meant was to move the `};` on line 1232 that closes the `OnLeftClick` lambda to after the local functions

Reply:
OK, I misunderstood -_-